### PR TITLE
Backport CFG simplification pass from XLA backend

### PR DESCRIPTION
**Pull request by vchuravy**
This backports the CFG simplifaction backend from the XLA backend.
It is not used by the pipeline, XLA and tools Cthulhu.jl want to use
it to present simpler IR to either humans or subsequent passes.

### CODE REVIEW

**Comment by Keno** [comment]
Supersedes #29544

**Review by Keno — APPROVED** [review]
Ok with me, but I wrote this code ;). FWIW, it does generally seem to work. I use it a whole bunch for various projects.

**Comment by JeffBezanson** [comment]
Don't we want to do some kind of CFG simplification by default eventually? I see a lot of messy IR, e.g. a chain of 10 adjacent gotos and the like.

**Comment by Keno** [comment]
I think that'll generally go away once we enable cfg simplifications I'm
compaction, since the round trip to CodeInfo for caching sort of acts like
a limited cfg simplify.

**Comment by vchuravy** [comment]
Ah I forgot about that branch and backported it straight from kf/tpu3.

**Comment by Keno** [comment]
You may want to diff that PR just to make sure I didn't have any bugfixes that I accidentally just applied there. That PR is definitely the newest version of that code.

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -983,14 +997,14 @@ end
 
 function finish_current_bb!(compact, active_bb, old_result_idx=compact.result_idx, unreachable=false)
     if compact.active_result_bb > length(compact.result_bbs)
-        @assert compact.bb_rename[active_bb] == 0
+        #@assert compact.bb_rename[active_bb] == 0
```
Comment (vchuravy) [inline base/compiler/ssair/ir.jl]: @Keno should this assert be re-enabled?

**Review by vchuravy — COMMENTED** [review]
Diff against kf/cfgsimp looks clean

File: base/compiler/ssair/ir.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/ssair/ir.jl]: It'd probably be correct to require this to be zero in both rename arrays, but I'm not entirely sure.

File: base/compiler/ssair/passes.jl [hunk 2]
```diff
@@ -1012,3 +1012,131 @@ function type_lift_pass!(ir::IRCode)
...
+        orig_bb = result_bbs[i]
+        preds = bbs[orig_bb].preds
+        map(preds) do pred
+            while merge_into[pred] != 0
+                pred = merge_into[pred]
+            end
+            bb_rename_succ[pred]
+        end
+    end
+    cresult_bbs = [BasicBlock(
```
Comment (vtjnash) [inline base/compiler/ssair/passes.jl]: used Typed comprehension to avoid potential performance issues from making a closure

**Pull request merged**